### ISSUE TITLE
keepalive for SSH connection during SCP Transfer

### ISSUE THREAD

**Issue by ameer131** [ktbyers/netmiko#1082]
hi @ktbyers ,
I am recently working on to automate SCP transfer for Juniper device.

If I am using your example here:
https://pynet.twb-tech.com/blog/automation/netmiko-scp.html

I will end up in two problem:
1) The Python process is killed - I fixed this by sizing the correct memory. Anyway, is there any baseline maximum memory that I can use? I am trying to minimize as much as resource possible.
2) Based on my research, for juniper device, it seems like the module tried to start a shell and use a shell command in order to get the detail of the filesystem in order to calculate disk space availability. My ID used to login into the device is not a rood ID, hence I will not be able to run a shell command. See below:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/app-root/lib/python3.5/site-packages/netmiko/scp_functions.py", line 70, in file_transfer
    if scp_transfer.check_file_exists():
  File "/opt/app-root/lib/python3.5/site-packages/netmiko/juniper/juniper.py", line 224, in check_file_exists
    return self._check_file_exists_unix(remote_cmd=remote_cmd)
  File "/opt/app-root/lib/python3.5/site-packages/netmiko/scp_handler.py", line 172, in _check_file_exists_unix
    self.ssh_ctl_chan._enter_shell()
  File "/opt/app-root/lib/python3.5/site-packages/netmiko/juniper/juniper.py", line 36, in _enter_shell
    return self.send_command('start shell sh', expect_string=r"[\$#]")
  File "/opt/app-root/lib/python3.5/site-packages/netmiko/base_connection.py", line 1188, in send_command
    search_pattern))
OSError: Search pattern never detected in send_command_expect: [\$#]
>>>

```
scp_handler.py:

```
    def _remote_space_available_unix(self, search_pattern=""):
        """Return space available on *nix system (BSD/Linux)."""
        self.ssh_ctl_chan._enter_shell()
        remote_cmd = "/bin/df -k {}".format(self.file_system)
        remote_output = self.ssh_ctl_chan.send_command(remote_cmd, expect_string=r"[\$#]")


```

Can we work on something to replace the above by maybe issuing below command:
`show system storage detail`

My workaround for this is to use the other way which is to open a SCP connection using below method:
1) Make connection to the device and named it as ssh_conn

```
source_file = 'myimage.tgz'
dest_fle = '/var/tmp/myimage.tgz'
ssh_conn_scp = netmiko.SCPConn(ssh_conn)
ssh_conn_scp.scp_transfer_file(source_file, dest_file)
ssh_conn_scp.close()

```

The transfer works great, I have verified the image and the md5 match. However, I noticed if it takes too long, the main ssh_conn will get disconnected and return me an error saying the socket is closed.
`Error Message: Socket is closed`

I believe that you have created is_alive function before in base_connection.py, my question is, does is_alive works when we are transferring the file over?

I am looking for possibility to use the above command and not closing the main ssh_conn. Reason being is that after I successfully copy the image, i will use send_command to verify md5 to ensure it is valid.

Thanks!

Regards,
Ameer.

**Comment by rwill-at-ncgov** [comment]
I have the same issue "Socket is closed" when attempting transfer of large Cisco ASR images. The transfer takes about an hour and I've not figured out how to tell netmiko to maintain the control session throughout. Thus, the file_transfer function is not useful. Is there some solution to this?

**Comment by ktbyers** [comment]
@rwill-at-ncgov You have to make sure that the VTY session for the SSH control channel doesn't timeout. So you need to execute the equivalent of `line vty 0 4 > exec-timeout 0`.

@ameer131 might have a different issue than this. I need to read more close about what he described.

**Comment by rwill-at-ncgov** [comment]
Thanks, Kirk. I was able to verify this works after modifying device for line vty > exec-timeout 120 0

Regards,

Rick

**Comment by ameer131** [comment]
hi @ktbyers ,

Did you have a chance to look into this matter?

At this moment, the workaround fix that I did is check prompt present, if its not found, it will perform a reconnection. it works great so far.

**Comment by ktbyers** [comment]
Is your issue the disconnect issue or that you don't have the root-shell access or both. The root-shell issue, is not having this access very common on Juniper.

On the disconnect issue, have you made sure your Netmiko-SSH session is not getting terminated by the Juniper device? There is nothing in Netmiko that would disconnect the session so the cause is likely either the Juniper device or an intermediate firewall.

is_alive() is probably not going to work (unless you are doing something special with concurrency) as you will be blocking by the secure copy call and the is_alive would not be executed until the secure copy is over.


Regards, Kirk

**Comment by daanvdsanden** [comment]
> Can we work on something to replace the above by maybe issuing below command:
> `show system storage detail`

I just created pull request #1297 so it is possible to copy files without shell access.

**Comment by carlmontanari** [comment]
I think this thread sounds like its mostly solved minus the socket timeout -- however I believe that is solved via  #1397, so I'm going to go ahead and close this out. Feel free to reopen if I read that wrong and the socket timeout doesnt sort the rest of this out!
